src/System.Net.HttpListener/tests/HttpListenerContextTests.cs

```diff
@@ -0,0 +1,179 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Net.Sockets;
+using System.Net.WebSockets;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.Net.Tests
+{
+    public class HttpListenerContextTests
+    {
+        public static bool IsNotWindows7 { get; } = !PlatformDetection.IsWindows7 && PlatformDetection.IsNotOneCoreUAP;
+
+        public static IEnumerable<object[]> SubProtocol_TestData()
+        {
+            yield return new object[] { new string[0], null };
+
+            yield return new object[] { new string[] { "MyProtocol1" }, null };
+            yield return new object[] { new string[] { "MyProtocol" }, "MyProtocol" };
+            yield return new object[] { new string[] { "MyProtocol" }, "myPROTOCOL" };
+
+            yield return new object[] { new string[] { "MyProtocol1", "MyProtocol2" }, null };
+            yield return new object[] { new string[] { "MyProtocol1", "MyProtocol2" }, "MyProtocol2" };
+        }
+
+        [ConditionalTheory(nameof(IsNotWindows7))]
+        [MemberData(nameof(SubProtocol_TestData))]
+        public async Task AcceptWebSocketAsync_ValidSubProtocol_Success(string[] clientProtocols, string serverProtocol)
+        {
+            await GetWebSocketContext((socket, context) =>
+            {
+                HttpListenerWebSocketContext socketContext = context.AcceptWebSocketAsync(serverProtocol).Result;
```

Comment:
I found that the tests hang if an assertion in the async lambda fails, but I was probably doing something wrong. I just pushed a commit that cleans up the tests to avoid using lambdas/callbacks altogether